### PR TITLE
Improve performance of sample! function

### PR DESCRIPTION
**Pull request by Tortar**
Seen that the implementation of add_newids! was improvable since some operations can be substitute with much better one, this is the code I used for the benchmark:

<details> <summary> code </summary>

```julia
using Agents, BenchmarkTools
using Agents: sample!, sample2!

mutable struct FakeAgent <: AbstractAgent
    id::Int
end

function fake_model(; nagents, scheduler)
    model = StandardABM(FakeAgent; scheduler)
    for i in 1:nagents
        add_agent!(model)
    end
    model
end

model = fake_model(; nagents = 1000, scheduler=Schedulers.fastest)

@benchmark sample!($model, 100000)
@benchmark sample2!($model, 100000)

@benchmark sample!($model, 10000)
@benchmark sample2!($model, 10000)

@benchmark sample!($model, 1000)
@benchmark sample2!($model, 1000)

@benchmark sample2!($model, 100)
@benchmark sample!($model, 100)

@benchmark sample!($model, 10)
@benchmark sample2!($model, 10)

@benchmark sample!($model, 1)
@benchmark sample2!($model, 1)
```

</details>

```julia
julia> @benchmark sample!($model, 100000)
BenchmarkTools.Trial: 3 samples with 1 evaluation.
 Range (min … max):  2.115 s …   2.196 s  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     2.167 s              ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.160 s ± 40.738 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                   █                   █  
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▁
  2.12 s         Histogram: frequency by time         2.2 s <

 Memory estimate: 18.11 MiB, allocs estimate: 146648.

julia> @benchmark sample2!($model, 100000)
BenchmarkTools.Trial: 150 samples with 1 evaluation.
 Range (min … max):  30.891 ms … 39.608 ms  ┊ GC (min … max): 0.00% … 14.33%
 Time  (median):     31.847 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   33.473 ms ±  2.762 ms  ┊ GC (mean ± σ):  4.66% ±  6.50%

    ▄█▅▇▇                                                      
  ▅▇█████▆▇▁▆▅▃▃▄▃▁▃▁▁▃▃▄▁▁▁▁▁▁▁▁▁▃▃▁▃▄▃▃▃▄▃▄▅▃▄▄▆▄▄▅▃▁▄▄▃▁▃▃ ▃
  30.9 ms         Histogram: frequency by time        39.5 ms <

 Memory estimate: 23.74 MiB, allocs estimate: 146228.

julia> @benchmark sample!($model, 10000)
BenchmarkTools.Trial: 231 samples with 1 evaluation.
 Range (min … max):  21.470 ms …  25.303 ms  ┊ GC (min … max): 0.00% … 6.80%
 Time  (median):     21.608 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   21.713 ms ± 395.993 μs  ┊ GC (mean ± σ):  0.25% ± 1.30%

    ▃█▄                                                         
  ▄█████▆▃▄▅▃▃▃▂▃▃▃▃▂▂▁▁▁▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▃▂▃ ▃
  21.5 ms         Histogram: frequency by time         23.5 ms <

 Memory estimate: 2.33 MiB, allocs estimate: 14316.

julia> @benchmark sample2!($model, 10000)
BenchmarkTools.Trial: 1899 samples with 1 evaluation.
 Range (min … max):  2.484 ms …   4.545 ms  ┊ GC (min … max): 0.00% … 39.28%
 Time  (median):     2.559 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.631 ms ± 342.597 μs  ┊ GC (mean ± σ):  2.46% ±  7.61%

  ▃▇█▅▃                                                     ▁  
  █████▇▅▅▆▇▇▆▃▃▁▁▃▅▄▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▃██ █
  2.48 ms      Histogram: log(frequency) by time      4.38 ms <

 Memory estimate: 2.67 MiB, allocs estimate: 14216.

julia> @benchmark sample!($model, 1000)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  381.873 μs …   2.228 ms  ┊ GC (min … max): 0.00% … 80.66%
 Time  (median):     405.041 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   412.388 μs ± 100.611 μs  ┊ GC (mean ± σ):  1.40% ±  4.65%

                ▁▂▄▄▆▇▇███▇▆▃▃                                   
  ▁▁▁▁▁▁▁▂▂▂▃▄▅▆███████████████▇▆▅▄▄▃▃▂▂▂▂▂▁▁▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃
  382 μs           Histogram: frequency by time          445 μs <

 Memory estimate: 248.50 KiB, allocs estimate: 1332.

julia> @benchmark sample2!($model, 1000)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  232.972 μs …   2.035 ms  ┊ GC (min … max): 0.00% … 83.72%
 Time  (median):     253.082 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   261.471 μs ± 110.073 μs  ┊ GC (mean ± σ):  2.69% ±  5.54%

               ▁▃▄▆█████▇▆▅▂▁                                    
  ▁▁▁▁▁▁▂▂▃▃▄▅▇███████████████▆▅▄▄▃▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃
  233 μs           Histogram: frequency by time          292 μs <

 Memory estimate: 273.97 KiB, allocs estimate: 1350.

julia> @benchmark sample2!($model, 100)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  19.969 μs …  2.679 ms  ┊ GC (min … max): 0.00% … 98.63%
 Time  (median):     26.622 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   28.048 μs ± 51.994 μs  ┊ GC (mean ± σ):  3.68% ±  1.97%

                  ▁▂▄▅▆▇▇▇▇▇█▆▅▃▄▂▁▁                           
  ▂▁▂▂▂▂▂▂▃▃▄▄▅▆▆█████████████████████▇▆▅▅▅▄▄▄▄▄▄▃▃▃▃▃▂▂▂▂▂▂▂ ▅
  20 μs           Histogram: frequency by time        35.3 μs <

 Memory estimate: 28.84 KiB, allocs estimate: 117.

julia> @benchmark sample!($model, 100)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  18.427 μs …  2.262 ms  ┊ GC (min … max): 0.00% … 96.74%
 Time  (median):     25.019 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   26.004 μs ± 37.846 μs  ┊ GC (mean ± σ):  2.48% ±  1.69%

                   ▂▃▃▄▅▆███▇▇█▆▆▄▃▃▃▁                         
  ▂▁▂▂▂▂▂▂▃▃▃▄▄▅▆▇▇████████████████████▇█▇▆▆▅▅▅▅▄▄▃▃▃▃▃▃▃▃▂▂▂ ▅
  18.4 μs         Histogram: frequency by time        32.5 μs <

 Memory estimate: 22.48 KiB, allocs estimate: 107.

julia> @benchmark sample!($model, 10)
BenchmarkTools.Trial: 10000 samples with 10 evaluations.
 Range (min … max):  1.663 μs … 288.778 μs  ┊ GC (min … max): 0.00% … 98.95%
 Time  (median):     2.472 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.623 μs ±   4.925 μs  ┊ GC (mean ± σ):  3.24% ±  1.71%

                  ▁▃▅▆▆▇█▆▅▅▃▃▂▁▁▁   ▁▁▂▁▁▁▁▁▂▁                
  ▂▁▁▂▂▂▂▃▃▃▄▄▅▆██████████████████▇██████████████▇▇▆▅▅▄▄▄▃▃▃▃ ▅
  1.66 μs         Histogram: frequency by time        3.34 μs <

 Memory estimate: 2.36 KiB, allocs estimate: 16.

julia> @benchmark sample2!($model, 10)
BenchmarkTools.Trial: 10000 samples with 9 evaluations.
 Range (min … max):  2.038 μs … 324.503 μs  ┊ GC (min … max): 0.00% … 98.82%
 Time  (median):     2.771 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.944 μs ±   5.516 μs  ┊ GC (mean ± σ):  3.23% ±  1.71%

              ▂▃▄▆▇▇█▇█▆▆▅▄▃▁▁    ▁ ▁▁▁▁▂  ▁                   
  ▁▁▂▂▂▂▃▄▅▆▇██████████████████████▇█████████▇▇▆▆▄▅▄▄▃▃▂▂▂▂▂▂ ▅
  2.04 μs         Histogram: frequency by time        3.78 μs <

 Memory estimate: 2.86 KiB, allocs estimate: 19.

julia> @benchmark sample!($model, 1)
BenchmarkTools.Trial: 10000 samples with 984 evaluations.
 Range (min … max):  54.883 ns …   3.120 μs  ┊ GC (min … max): 0.00% … 98.01%
 Time  (median):     57.562 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   67.018 ns ± 117.717 ns  ┊ GC (mean ± σ):  7.00% ±  3.91%

     ▃█▃                                                        
  ▁▂▄███▆▄▃▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▂▃▄▄▃▂▂▁▁▁▁▁▁▁▁▁ ▂
  54.9 ns         Histogram: frequency by time         84.3 ns <

 Memory estimate: 128 bytes, allocs estimate: 2.

julia> @benchmark sample2!($model, 1)
BenchmarkTools.Trial: 10000 samples with 852 evaluations.
 Range (min … max):  139.191 ns …   4.947 μs  ┊ GC (min … max):  0.00% … 95.33%
 Time  (median):     147.000 ns               ┊ GC (median):     0.00%
 Time  (mean ± σ):   188.013 ns ± 374.215 ns  ┊ GC (mean ± σ):  16.95% ±  8.18%

  ▂▆▇█▇▆▅▄▄▃▃▃▃▂▁                                      ▁▂▂▂▂    ▂
  ████████████████▇▇▆▆▇▇▆▆▆▅▄▄▅▅▆▆▆▅▅▄▅▄▄▅▄▅▅▅▄▅▇▇▅▆▅▅▆███████▇ █
  139 ns        Histogram: log(frequency) by time        256 ns <

 Memory estimate: 672 bytes, allocs estimate: 6.
```

As you can see this new implementation is already faster with a sample of 100 agents, and it become faster and faster more the agents number increases, at 100000 agents this new function is 70x faster! This will have a direct impact on the `wright-fisher` model example where a similar speed-up is expected!

### CODE REVIEW

File: src/simulations/sample.jl [hunk 1]
```diff
@@ -35,16 +35,18 @@ function sample!(
...
-"Used in sample!"
+#Used in sample!
 function add_newids!(model, org_ids, newids)
     n = nextid(model)
+    count_newids = counter(newids)
     for id in org_ids
-        if !in(id, newids)
-            kill_agent!(model.agents[id], model)
+        noccurances = count_newids[id]
+        agent = model.agents[id]
```
Comment (Datseris) [inline src/simulations/sample.jl:44]:
```suggestion
        agent = model[id]
```

File: src/simulations/sample.jl [hunk 2]
```diff
@@ -35,16 +35,18 @@ function sample!(
     add_newids!(model, org_ids, newids)
 end
 
-"Used in sample!"
+#Used in sample!
 function add_newids!(model, org_ids, newids)
     n = nextid(model)
+    count_newids = counter(newids)
```
Comment (Datseris) [inline src/simulations/sample.jl:41]: Where does `counter` come from...?

Comment (Tortar) [reply src/simulations/sample.jl:41]: it comes from `DataStructures.jl`

Comment (Datseris) [reply src/simulations/sample.jl:41]: aaaah maybe write a comment about htis in the source...?

Comment (Tortar) [reply src/simulations/sample.jl:41]: done!

**Comment by Tortar** [comment]
I'm running the test for a sample of 1 million and the previous implementation hasn't returned anything yer, probably a 1000x at least :D

**Review by Datseris — APPROVED** [review]
awesome, thanks!

Same story as before: add a changelog entry and merge!

**Pull request merged**